lib/edda-server/src/compressor.rs

```diff
@@ -0,0 +1,422 @@
+    workspace_snapshot::graph::detector::Update,
+};
+use edda_core::api_types::{
+    RequestId,
+    compressor_rebuild_request::{
+        CompressorRebuildRequest,
+        CompressorRebuildRequestVCurrent,
+    },
+    compressor_update_request::{
+        CompressorUpdateRequest,
+        CompressorUpdateRequestV1,
+        CompressorUpdateRequestVCurrent,
+    },
+    rebuild_request,
+};
+use futures::{
+    FutureExt,
+    StreamExt,
+};
+use naxum::{
+    MessageHead,
+    extract::{
+        FromMessage,
+        FromMessageRaw,
+    },
+};
+use pin_project_lite::pin_project;
+use si_data_nats::async_nats::jetstream::{
+    self,
+    Message,
+    consumer::{
+        self,
+        StreamError,
+        push,
+    },
+    stream::ConsumerError,
+};
+use si_events::workspace_snapshot::Change;
+use si_id::EntityId;
+use si_layer_cache::LayerDbError;
+use telemetry::prelude::*;
+use thiserror::Error;
+
+use crate::extract::{
+    ApiTypesNegotiate,
+    EddaRequestKind,
+};
+
+#[remain::sorted]
+#[derive(Debug, Error)]
+pub enum CompressorError {
+    #[error("error creating consumer: {0}")]
+    ConsumerCreate(#[source] ConsumerError),
+    #[error("error building dal context: {0}")]
+    DalContextBuild(#[source] TransactionsError),
+    #[error("invalid edda request kind: compressor rebuild")]
+    InvalidEddaRequestKindCompressorRebuild,
+    #[error("invalid edda request kind: compressor update")]
+    InvalidEddaRequestKindCompressorUpdate,
+    #[error("layer db error: {0}")]
+    LayerDb(#[from] LayerDbError),
+    #[error("error while subscribing for messages: {0}")]
+    Subscribe(#[source] StreamError),
+}
+
+type Result<T> = std::result::Result<T, CompressorError>;
+
+#[remain::sorted]
+#[derive(Debug, Clone)]
+enum State {
+    Draining(CompressorUpdateRequestVCurrent),
+    DrainingToRebuild(CompressorRebuildRequestVCurrent),
+    Inactive,
+}
+
+pin_project! {
+    /// A stream that wraps an inner stream with the ability to "compress" requests.
+    pub struct Compressor {
+        incoming: consumer::push::Ordered,
+        state: State,
+        ctx_builder: DalContextBuilder,
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+        ctx: DalContext,
+    }
+}
+
+impl Compressor {
+    pub async fn new(
+        requests_stream: &jetstream::stream::Stream,
+        consumer_config: push::OrderedConfig,
+        ctx_builder: DalContextBuilder,
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+    ) -> Result<Self> {
+        let incoming = requests_stream
+            .create_consumer(consumer_config)
+            .await
+            .map_err(CompressorError::ConsumerCreate)?
+            .messages()
+            .await
+            .map_err(CompressorError::Subscribe)?;
+
+        let ctx = ctx_builder
+            .build_for_change_set_as_system(workspace_id, change_set_id, None)
+            .await
+            .map_err(CompressorError::DalContextBuild)?;
```

Comment:
This is really only used for layer db. In theory we should also be using it for checking if the snapshot has moved, but the "is this contiguous" request compression logic should be fine. I think if we don't need a full context, we should consider just giving the compressor its own layer db client.